akka-stream/src/main/scala/akka/stream/impl/UnfoldResourceSourceAsync.scala

```diff
@@ -27,83 +26,86 @@ import scala.util.control.NonFatal
+    private implicit def ec = ActorMaterializerHelper.downcast(materializer).system.dispatcher
+    private var state: Option[S] = None
+
+    private val createdCallback = getAsyncCallback[Try[S]] {
+      case Success(resource) ⇒
+        state = Some(resource)
+        if (isAvailable(out)) onPull()
+      case Failure(t) ⇒ failStage(t)
+    }.invoke _
 
-    val errorHandler: PartialFunction[Throwable, Unit] = {
+    private val errorHandler: PartialFunction[Throwable, Unit] = {
       case NonFatal(ex) ⇒ decider(ex) match {
         case Supervision.Stop ⇒
-          onResourceReady(close(_))
           failStage(ex)
-        case Supervision.Restart ⇒ restartState()
+        case Supervision.Restart ⇒ restartResource()
         case Supervision.Resume  ⇒ onPull()
       }
     }
 
-    val readCallback = getAsyncCallback[Try[Option[T]]] {
-      case scala.util.Success(data) ⇒ data match {
+    private val readCallback = getAsyncCallback[Try[Option[T]]] {
+      case Success(data) ⇒ data match {
         case Some(d) ⇒ push(out, d)
-        case None    ⇒ closeStage()
+        case None ⇒
+          // end of resource reached, lets close it
+          state match {
+            case Some(resource) ⇒
+              close(resource).onComplete(getAsyncCallback[Try[Done]] {
+                case Success(Done) ⇒ completeStage()
+                case Failure(ex)   ⇒ failStage(ex)
+              }.invoke)
+              state = None
+
+            case None ⇒
+              // cannot happen, but for good measure
+              throw new IllegalStateException("Reached end of data but there is no connection")
+          }
       }
-      case scala.util.Failure(t) ⇒ errorHandler(t)
+      case Failure(t) ⇒ errorHandler(t)
     }.invoke _
 
-    final override def onPull(): Unit =
-      onResourceReady { resource ⇒
-        try { readData(resource).onComplete(readCallback) } catch errorHandler
+    override def preStart(): Unit = createResource()
+
+    override def onPull(): Unit =
+      state match {
+        case Some(resource) ⇒
+          try {
+            readData(resource).onComplete(readCallback)
+          } catch errorHandler
+        case None ⇒
+        // we got a pull but there is no connection, we are either
+        // currently connecting/reconnecting and then the read will be triggered when connection completes,
+        // or shutting down and then the pull does not matter anyway
       }
 
-    override def onDownstreamFinish(): Unit = closeStage()
-
-    private def closeAndThen(f: () ⇒ Unit): Unit = {
-      setKeepGoing(true)
-      val closedCallback = getAsyncCallback[Try[Done]] {
-        case scala.util.Success(_) ⇒
-          open = false
-          f()
-        case scala.util.Failure(t) ⇒
-          open = false
-          failStage(t)
-      }
+    override def postStop(): Unit = {
+      if (state.isDefined) close(state.get)
```

Comment:
We don't wait for close here, this should be the actual bug fix guaranteeing that we always close the stage (old logic did incorrect dance with async callback inside `closeStage`/`closeAndThen`)